pkg/parser/grok/cache.go

```diff
@@ -0,0 +1,212 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// copied from https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/pkg/stanza/operator/parser/regex/cache.go
+
+package grok
+
+import (
+	"math"
+	"sync"
+	"time"
+
+	"go.uber.org/atomic"
+)
+
+// cache allows operators to cache a value and look it up later
+type cache interface {
+	get(key string) interface{}
+	add(key string, data interface{}) bool
+	copy() map[string]interface{}
+	maxSize() uint16
+}
+
+// newMemoryCache takes a cache size and a limiter interval and
+// returns a new memory backed cache
+func newMemoryCache(maxSize uint16, interval uint64) *memoryCache {
+	// start throttling when cache turnover is above 100%
+	limit := uint64(maxSize) + 1
+
+	return &memoryCache{
+		cache:   make(map[string]interface{}),
+		keys:    make(chan string, maxSize),
+		limiter: newStartedAtomicLimiter(limit, interval),
+	}
+}
+
+// memoryCache is an in memory cache of items with a pre defined
+// max size. Memory's underlying storage is a map[string]item
+// and does not perform any manipulation of the data. Memory
+// is designed to be as fast as possible while being thread safe.
+// When the cache is full, new items will evict the oldest
+// item using a FIFO style queue.
```

Comment:
Any reason why this is used here instead of the LRU kind cache? Say there is an item which is added first to the queue, but it's also the most commonly used one for all the log messages, in that kicking that out in favour of some new item doesn't seem wise.

Reply:
Well while building this parser, I took the same cache that was available for regex. But your question is legit.

Can create a new issue and we can have a discussion over there.

Reply:
https://github.com/SigNoz/signoz-otel-collector/issues/71